fix build and tests on Solaris 10

wiggin15 opened the pull request.

Commit:
Fix get mtu on Solaris 10

Commit:
swap -k is not supported in Solaris 10

Commit:
fix test on Solaris

Commit:
fix test on Solaris

os.path.exists will fail if the link is broken. see comment about errno.ENOENT
in memory_maps in _pssunos.py

Commit:
fix test on Solaris

Commit:
fix test on Solaris

wiggin15 commented:
Note: there are still 3 tests that fail on Solaris 10, 2 of them because of the issue described in #517.

giampaolo commented:
Hey, I will travel to NYC soon so I won't time to look into this for a while (I hope not too much). Please be patient.

jliverman commented:
We experienced similar issues (and produced a similar patch) as lymanepp in #517.

We also found that the variable PSUTIL_VERSION was not defined and so we supplied it in the _psutil_sunos.h header file (resulting in a one-line patch). We just gave it a value of 221 for 2.2.1, the version from https://pypi.python.org/simple/ for easy_install.

For the most up-to-date version of psutil (here on github) there seemed to be additional issues. Namely, the function psutil_net_if_stats seems to rely on a struct typedef called ifreq from net/if.h. However, our versions of Solaris 10 specifically say:

    /*
     * OBSOLETE: Replaced by struct lifreq. Supported for compatibility.
     *
     * Interface request structure used for socket
     * ioctl's.  All interface ioctl's must have parameter
     * definitions which begin with ifr_name.  The
     * remainder may be interface specific.
     */
    struct  ifreq {
    ...

within the net/if.h header file.

Also, the ifreq struct seems to be missing a member called "ifr_mtu" that is referenced in the _psutil_sunos.c code. The suggested replacement (i.e. struct typedef lifreq) has a member called "lifru_mtu" that seems like the likely replacement. However, I just hard-coded the value to 0 to get everything to compile as I did not feel comfortable working with the types or required casts, if any (ifreq relies on standard types like int, short and char while lifreq relies on derived types like uint_t and uint64_t). 

At this time we're getting by with our little hacks. But we're hoping that the maintainer can address them soon. We have a fairly critical application that depends on it. But alas, we are not exactly "C people" if you know what I mean...

wiggin15 commented:
@jliverman The lifreq issue was reported in #609, the fix is in this pull request.
@giampaolo any chance to review this and get this merged? Thanks.

giampaolo commented:
I'm still in US, still having very few time to look at anything other than work (sigh!). Sorry.

wiggin15 commented:
Hi. I rebased and updated this pull request. Can you please review it?
Note that merging this pull request will also resolve #609, #620 and #621.
Also note that there are still 12 (!) tests that fail on Solaris 10 now, which I didn't look into yet (most are new tests that didn't exist when I first opened this pull request).

Commit:
Add ifaddrs.c for Solaris 10

ifaddrs doesn't exist on this platform, so I added an
implementation of it for use in _psutil_posix.c

wiggin15 commented:
I updated the branch, with support for net_if_addrs, and now we're down to 9 failing tests. I found and fixed the problem with 4 of the failing tests but the fixes are not Solaris specific so I will open a separate pull request for them. 3 tests fail because of the issue reported in #517. I'm not sure about the other 2.

giampaolo commented:
what's the test output?
Il 05/set/2015 11:13, "wiggin15" <notifications@github.com> ha scritto:

> I updated the branch, with support for net_if_addrs, and now we're down to
> 9 failing tests. I found and fixed the problem with 4 of the failing tests
> but the fixes are not Solaris specific so I will open a separate pull
> request for them. 3 tests fail because of the issue reported in #517
> <https://github.com/giampaolo/psutil/issues/517>. I'm not sure about the
> other 2.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/giampaolo/psutil/pull/610#issuecomment-137935458>.
>

wiggin15 commented:
I uploaded the output here: https://gist.github.com/wiggin15/ae62963b33716da3a503

I created this branch: https://github.com/Infinidat/psutil/commits/test_fixes
to fix 4 of the tests (test_cpu_count and the 3 TestUnicode tests). I'm working on making it pass the CI.
The ones with "couldn't find any network interface" are #517 related.
test_connections and test_prog_w_funky_name are left.

giampaolo commented on test/test_psutil.py:
```diff
@@ -1385,7 +1385,10 @@ def test_create_time(self):
         terminal = psutil.Process().terminal()
         if sys.stdin.isatty():
-            self.assertEqual(terminal, sh('tty'))
+            tty = sh('tty')
```
you could simply use `tty = os.path.realpath(sh('tty'))`

giampaolo commented on psutil/arch/solaris10/ifaddrs.c:
```diff
@@ -0,0 +1,117 @@
+/* Refrences:
```
Please put this file under:

`psutil/arch/solaris/v10/ifaddrs.c`

giampaolo replied:
Also please use 4 spaces for indentation.

giampaolo commented on psutil/arch/solaris10/ifaddrs.h:
```diff
@@ -0,0 +1,26 @@
+/* Reference: https://lists.samba.org/archive/samba-technical/2009-February/063079.html */
```
Please put this file under:

`psutil/arch/solaris/v10/ifaddrs.h`

giampaolo replied:
Also please use 4 spaces for indentation.

giampaolo commented on setup.py:
```diff
@@ -64,13 +65,15 @@ def write(self, s):
+        posix_extension.libraries.append('socket')
+        if platform.release() == '5.10':
+            posix_extension.sources.append('psutil/arch/solaris10/ifaddrs.c')
+            posix_extension.define_macros.append(('_SUNOS10', 1))
```
Rename this to `psutil_SUNOS10`

giampaolo commented on psutil/_psutil_posix.c:
```diff
@@ -13,9 +13,14 @@
-#include <ifaddrs.h>
 #include <net/if.h>
 
+#ifdef _SUNOS10
```
Rename this to `PSUTIL_SUNOS10`

giampaolo commented:
Good work! I added some comments. Please also do not forget to update HISTORY.rst and CREDITS files.

Commit:
Code review fixes

wiggin15 commented:
Thanks for the review. I fixed your comments.
